File: Solution/15.3-sum.go
package LeetCode

import "sort"

/*
 * @lc app=leetcode id=15 lang=golang
 *
 * [15] 3Sum
 */

// @lc code=start
func threeSum(nums []int) [][]int {
	// sort the input array in ascending order
	sort.Ints(nums)
	var result [][]int
	var length int = len(nums) - 1
	var target int
	var sum int
	for start := 0; start < length-1; start++ {
		// skip duplicate elements to avoid duplicates in the result
		if start > 0 && nums[start] == nums[start-1] {
			continue
		}
		// set the target value as the negative of the current element
		target = -nums[start]
		// use two pointers to find two elements that sum up to the target value
		for left, right := start+1, length; left < right; {
			sum = nums[left] + nums[right]
			if sum < target {
				// move the left pointer to the right to increase the sum
				left++
				continue
			}
			if sum > target {
				// move the right pointer to the left to decrease the sum
				right--
				continue
			}
			// found a triplet that sums up to zero, add it to the result
			result = append(result, []int{nums[start], nums[left], nums[right]})
			// skip duplicate elements to avoid duplicates in the result
			for left < right && nums[left] == nums[left+1] {
				left++
			}
			for left < right && nums[right] == nums[right-1] {
				right--
			}
			// move both pointers to find other potential triplets
			left++
			right--
		}
	}
	return result
}

// @lc code=end

File: 15.3-sum.py
#
# @lc app=leetcode id=15 lang=python
#
# [15] 3Sum
#


# @lc code=start
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        result = []
        length = len(nums) - 1
        target = 0
        sum_val = 0

        for start in range(length - 1):
            if start > 0 and nums[start] == nums[start - 1]:
                continue
            target = -nums[start]
            left, right = start + 1, length
            while left < right:
                sum_val = nums[left] + nums[right]
                if sum_val < target:
                    left += 1
                    continue
                if sum_val > target:
                    right -= 1
                    continue
                result.append([nums[start], nums[left], nums[right]])
                while left < right and nums[left] == nums[left + 1]:
                    left += 1
                while left < right and nums[right] == nums[right - 1]:
                    right -= 1
                left += 1
                right -= 1

        return result


# @lc code=end
